Sources/Testing/SourceAttribution/CustomTestStringConvertible.swift

```diff
@@ -0,0 +1,154 @@
+//
+
+/// A protocol describing types with a custom string representation when
+/// presented as part of a test's output.
+///
+/// Values whose types conform to this protocol use it to describe themselves
+/// when they are present as part of the output of a test. For example, this
+/// protocol affects the display of values that are passed as arguments to test
+/// functions or that are elements of an expectation failure.
+///
+/// By default, the testing library converts values to strings using
+/// `String(describing:)`. The resulting string may be inappropriate for some
+/// types and their values. If the type of the value is made to conform to
+/// ``CustomTestStringConvertible``, then the value of its ``testDescription``
+/// property will be used instead.
+///
+/// For example, consider the following type:
+///
+/// ```swift
+/// enum Food {
+///   case paella
+///   case oden
+///   case ragu
+///   ...
+/// }
+/// ```
+///
+/// If an array of cases from this enumeration is passed to a parameterized test
+/// function:
+///
+/// ```swift
+/// @Test(arguments: [.paella, .oden, .ragu])
+/// func isDelicious(_ food: Food) { ... }
+/// ```
+///
+/// Then the values in the array need to be presented in the test output, but
+/// the default description of a value may not be adequately descriptive:
+///
+/// ```
+/// ◇ Passing argument food → .paella to isDelicious(\_:)
+/// ◇ Passing argument food → .oden to isDelicious(\_:)
+/// ◇ Passing argument food → .ragu to isDelicious(\_:)
+/// ```
+///
+/// By adopting ``CustomTestStringConvertible``, customized descriptions can be
+/// included:
+///
+/// ```swift
+/// extension Food: CustomTestStringConvertible {
+///   var testDescription: String {
+///     switch self {
+///     case .paella:
+///       "paella valenciana"
+///     case .oden:
+///       "おでん"
+///     case .ragu:
+///       "ragù alla bolognese"
+///     }
+///   }
+/// }
+/// ```
+///
+/// The presentation of these values will then reflect the value of the
+/// ``testDescription`` property:
+///
+/// ```
+/// ◇ Passing argument food → paella valenciana to isDelicious(\_:)
+/// ◇ Passing argument food → おでん to isDelicious(\_:)
+/// ◇ Passing argument food → ragù alla bolognese to isDelicious(\_:)
+/// ```
+///
+/// ## See Also
+///
+/// - ``Swift/String/init(describingForTest:)``
+public protocol CustomTestStringConvertible {
+  /// A description of this instance to use when presenting it in a test's
+  /// output.
+  ///
+  /// Do not use this property directly. To get the test description of a value,
+  /// use ``Swift/String/init(describingForTest:)``.
+  var testDescription: String { get }
+}
+
+extension String {
+  /// Initialize this instance so that it can be presented in a test's output.
+  ///
+  /// - Parameters:
+  ///   - value: The value to describe.
+  ///
+  /// ## See Also
+  ///
+  /// - ``CustomTestStringConvertible``
+  public init<T>(describingForTest value: T) {
```

Comment:
Not `some Any`?

Reply:
I wonder if we _need_ this to be `public`? The testing library certainly needs to call it, but do users? I can imagine scenarios where it might become useful but it might be best for this to be internal until we find a concrete need.

Reply:
It's parallel to `String(describing:)` and `CustomStringConvertible`. I _think_ it makes sense to expose both at the same visibility (whether that's `public`, `@_spi`, or something else.)